### ISSUE TITLE
Ubuntu 20.04, PHP7.4, Apache2+HTTPS Challenges?

### ISSUE THREAD

**Issue by cknight725** [phpvirtualbox/phpvirtualbox#237]
Is anyone else having (or have tried) using phpvirtualbox in the above config?

I get the lovely "An unknown PHP error occurred. This is most likely a syntax error in config.php in phpVirtualBox's folder. The most common errors are an unclosed quote or a missing semicolon in a configuration item that has been entered (e.g. location, username, or password)." Error ...

Navigating to config.php comes back clean, I have done
```php -l config.php ```
which also comes back clean --"No syntax errors detected in ./config.php" .

Tried:
```
apt reinstall php7.4 php7.4-mysql php7.4-soap
service apache2 restart
```
Which did not fix the problem.  I'm not finding logging that is helpful in either /var/log/* and the only possibly useful (likely not) hint is a browser console error --
``` 
jquery-1.11.2.min.js:4 [Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
```

apache error.log traces the following on page load:
```
Tue Jun 16 13:21:03.706294 2020] [ssl:info] [pid 166293] [client 192.168.1.9:59234] AH01964: Connection to child 3 established (server officepc.domain.fqdn:443)
[Tue Jun 16 13:21:03.706602 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(2372): [client 192.168.1.9:59234] AH02043: SSL virtual host for servername officepc.domain.fqdn found
[Tue Jun 16 13:21:03.706617 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(2372): [client 192.168.1.9:59234] AH02043: SSL virtual host for servername officepc.domain.fqdn found
[Tue Jun 16 13:21:03.706623 2020] [core:debug] [pid 166293] protocol.c(2313): [client 192.168.1.9:59234] AH03155: select protocol from , choices=h2,http/1.1 for server officepc.domain.fqdn
[Tue Jun 16 13:21:03.708868 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(2254): [client 192.168.1.9:59234] AH02041: Protocol: TLSv1.3, Cipher: TLS_AES_256_GCM_SHA384 (256/256 bits)
[Tue Jun 16 13:21:03.708976 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(493): AH00831: socache_shmcb_store (0xe9 -> subcache 9)
[Tue Jun 16 13:21:03.708995 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(847): AH00847: insert happened at idx=0, data=(0:32)
[Tue Jun 16 13:21:03.709002 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(850): AH00848: finished insert, subcache: idx_pos/idx_used=0/1, data_pos/data_used=0/228
[Tue Jun 16 13:21:03.709008 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(515): AH00834: leaving socache_shmcb_store successfully
[Tue Jun 16 13:21:03.709105 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(493): AH00831: socache_shmcb_store (0x3e -> subcache 30)
[Tue Jun 16 13:21:03.709120 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(847): AH00847: insert happened at idx=0, data=(0:32)
[Tue Jun 16 13:21:03.709126 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(850): AH00848: finished insert, subcache: idx_pos/idx_used=0/1, data_pos/data_used=0/228
[Tue Jun 16 13:21:03.709132 2020] [socache_shmcb:debug] [pid 166293] mod_socache_shmcb.c(515): AH00834: leaving socache_shmcb_store successfully
[Tue Jun 16 13:21:03.709269 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(415): [client 192.168.1.9:59234] AH02034: Initial (No.1) HTTPS request received for child 3 (server officepc.domain.fqdn:443)
[Tue Jun 16 13:21:03.709331 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of Require all granted: granted
[Tue Jun 16 13:21:03.709342 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of <RequireAny>: granted
[Tue Jun 16 13:21:03.709405 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of Require all granted: granted
[Tue Jun 16 13:21:03.709422 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of <RequireAny>: granted
[Tue Jun 16 13:21:03.710302 2020] [deflate:debug] [pid 166293] mod_deflate.c(854): [client 192.168.1.9:59234] AH01384: Zlib: Compressed 19864 to 5227 : URL /phpvirtualbox/index.html
[Tue Jun 16 13:21:03.877231 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(415): [client 192.168.1.9:59234] AH02034: Subsequent (No.2) HTTPS request received for child 3 (server officepc.domain.fqdn:443), referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.877284 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of Require all granted: granted, referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.877306 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of <RequireAny>: granted, referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.878327 2020] [deflate:debug] [pid 166293] mod_deflate.c(854): [client 192.168.1.9:59234] AH01384: Zlib: Compressed 4802 to 1243 : URL /phpvirtualbox/endpoints/language.php, referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.918226 2020] [ssl:debug] [pid 166293] ssl_engine_kernel.c(415): [client 192.168.1.9:59234] AH02034: Subsequent (No.3) HTTPS request received for child 3 (server officepc.domain.fqdn:443), referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.918285 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of Require all granted: granted, referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:03.918295 2020] [authz_core:debug] [pid 166293] mod_authz_core.c(817): [client 192.168.1.9:59234] AH01626: authorization result of <RequireAny>: granted, referer: https://officepc.domain.fqdn/phpvirtualbox/
[Tue Jun 16 13:21:08.920640 2020] [ssl:debug] [pid 166293] ssl_engine_io.c(1102): [client 192.168.1.9:59234] AH02001: Connection closed to child 3 with standard shutdown (server officepc.domain.fqdn:443)
```

**Comment by cknight725** [comment]
Update -- tracked the "problem" to Content Security Policy -- so the question is -- what are the right CSP directives for a secure webserver and phpVirtualBox?

This was the setting in apache2.conf:

```Header set Content-Security-Policy "default-src 'none'; script-src 'self' 'unsafe-inline' https://*.domain.fqdn; connect-src 'self'; img-src 'self' data:; style-src 'unsafe-inline'; style-src-elem 'unsafe-inline' https://*.domain.fqdn/ https://*.googleapis.com/; base-uri 'self';form-action 'self'; font-src https://*.domain.fqdn/ https://fonts.gstatic.com/"```

**Comment by cknight725** [comment]
Found the problem.  In order for CSP to allow phpVirtualBox you must set script-src 'unsafe-eval' -- 

`Header set Content-Security-Policy "default-src 'none'; script-src 'self' **'unsafe-inline' 'unsafe-eval'** *.domain.fqdn www.google-analytics.com; connect-src 'self' *.domain.fqdn; img-src 'self' data: www.google-analytics.com; style-src 'self' 'unsafe-inline' *.domain.fqdn; style-src-elem 'self' 'unsafe-inline' *.domain.fqdn https://*.googleapis.com; base-uri 'self'; form-action 'self'; font-src 'self' *.domain.fqdn https://fonts.gstatic.com"`

Basically phpVirtualBox isn't particularly secure in 2020 - probably not a big deal given its prevalence in non-prod and lab environments -- but its probably a bad idea to put it in prod on the Internet (or in the Cloud without ACLs).

**Comment by h6w** [comment]
Hi, yeah.  Thanks.  I actually would like to keep this open.  While I agree with your conclusion, and I don't have phpvirtualbox publicly available anywhere myself, there are situations where it would still be secure to do so and your bug would affect it.  e.g. Source-based-locking, GRE networks, or VPNs.  So I'm reopening this to remind me to fix it.

**Comment by atx49559** [comment]
By my case, this command can fix that.
apt install php7.3-xml

Debian 10
Apache 2.4.3.8
php v7.3.14

https://sourceforge.net/p/phpvirtualbox/discussion/help/thread/c566d48f7a/
